src/perf/bin/drvmain.cpp

```diff
@@ -585,9 +606,16 @@ QUIC_THREAD_CALLBACK(PerformanceWaitForStopThreadCb, Context)
         return;
     }
 
-    WdfRequestUnmarkCancelable(Request);
+    QuicLockAcquire(&Client->CleanupLock);
```

Comment:
I still don't like holding a lock while you call into here. You should be able to achieve synchronization without doing this.

Reply:
The call to WdfRequestUnmarkCancelable basically needs to happen with some sort of synchronize with setting the variable.

According to the docs, normally you use the Wdf automatic synchronization for this, which calls the whole cancel function under a lock, but I don't want to enable that. But this function is designed to be called under a lock, because it is 100% a race condition if you don't have it.

Reply:
I'm still not convinced but I haven't had a time to look in depth here.

Reply:
God this stuff is complicated. I was going to propose a different solution that didn't require a lock to be held while you call, but then read the docs more and realized it wouldn't work. Assuming you've tested canceling and this solution generally works, I'm fine with it for now. (it might not solve all edge conditions, but it's just tool/test code, so I don't care too much).